# 1. 데이터 로드

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import glob
from datetime import datetime as dt
from tqdm import tqdm

import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

train = pd.read_csv( 'input/funda_train.csv' )
submit = pd.read_csv( 'input/submission.csv' )


# 데이터형태 확인
train.head()

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   
3         0        3            a      2016-06-01           20:22   
4         0        4            c      2016-06-02           11:06   

   installment_term region type_of_business       amount  
0                 0    NaN           기타 미용업  1857.142857  
1                 0    NaN           기타 미용업   857.142857  
2                 0    NaN           기타 미용업  2000.000000  
3                 0    NaN           기타 미용업  7857.142857  
4                 0    NaN           기타 미용업  2000.000000

In [2]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6556613 entries, 0 to 6556612
Data columns (total 9 columns):
store_id            int64
card_id             int64
card_company        object
transacted_date     object
transacted_time     object
installment_term    int64
region              object
type_of_business    object
amount              float64
dtypes: float64(1), int64(3), object(5)
memory usage: 450.2+ MB


## 1.1 전역에서 활용할 함수들을 필요할 때 마다 정의

In [3]:
# DataFrame 의 컬럼에서 max 값 확인
def get_max( df, column_key ) :
    return df[column_key][df[column_key].idxmax()]

def get_min( df, column_key ) :
    return df[column_key][df[column_key].idxmin()]

## 1.2 데이터 확인
### 1.2.1 유니크 값 확인

In [4]:
# column 별 유니크 값 개수 확인
for key in train.columns :
    print( '{}: {}'.format( key, len( train[key].unique() ) ) )

store_id: 1967
card_id: 3950001
card_company: 8
transacted_date: 1003
transacted_time: 1440
installment_term: 34
region: 181
type_of_business: 146
amount: 30551


<b>[의미있는 값]</b><br>
1. 상점은 1967개다
 - 상점별 매출액을 종합할 수 있다.
2. 카드사는 총 8개다.
 - 분별력있는 판단은 줄 수 없지만 구분값으론 의미가 있을 수도 있다.
3. 181개 지역의 매출이다
 - 지역별 매출액에 의미를 줄 수있는지 확인해볼 수 있다.
4. 업종은 146개다
 - 계절에 따른 업종별 매출액이 의미를 줄 수 있는지 확인해볼 수 있다.

### 1.2.2 매출액 확인

In [5]:
# 매출액 최대 최소값 확인
print( get_min( train, 'amount' ) )
print( get_max( train, 'amount' ) )

-2771428.5714285714
5571428.571428572


In [6]:
# 음수인 매출액 확인
train[train['amount']<0]

store_id  card_id card_company transacted_date transacted_time  \
41              0       40            a      2016-06-10           17:26   
347             0      285            a      2016-08-04           17:52   
731             0      473            g      2016-10-17           10:32   
831             0      230            b      2016-11-03           15:36   
944             0      138            a      2016-11-28           13:21   
1953            0      829            a      2017-07-10           17:19   
2659            0     1182            e      2017-12-22           19:22   
2734            0     1080            g      2018-01-10           17:27   
3256            0      626            g      2018-05-13           18:16   
3544            0     1401            a      2018-07-16           14:22   
3700            0      889            a      2018-08-20           17:34   
3794            0     1537            a      2018-09-16           16:18   
7112            1     2992            b      2018-03-23           14:29   
7309            1     2702            a      2018-05-17           16:17   
7391            1     2922            f      2018-06-14           15:43   
7392            1     2922            f      2018-06-14           15:43   
7686            1     3206            b      2018-09-12           15:19   
8077            1     3331            b      2019-01-17           12:14   
8130            1     3382            b      2019-01-25           14:03   
8169            1     3401            a      2019-02-01           13:02   
8173            1     3402            b      2019-02-08           12:25   
8181            1     3191            e      2019-02-08           14:59   
8317            2     3579            b      2016-11-07           13:41   
8362            2     3613            c      2017-03-10           16:31   
8370            2     3617            f      2017-03-13           17:39   
8374            2     3613            c      2017-03-21           16:58   
8576            2     3762            b      2018-04-03           19:40   
8583            2     3765            f      2018-04-07           17:00   
8589            2     3768            f      2018-05-17           11:28   
8591            2     3768            f      2018-05-17           15:51   
...           ...      ...          ...             ...             ...   
6552013      2136  4661163            b      2018-02-05           21:54   
6552094      2136  4661212            d      2018-02-13           00:25   
6552366      2136  4660671            h      2018-03-04           00:10   
6553150      2136  4661822            d      2018-05-07           01:51   
6553305      2136  4661913            d      2018-05-20           00:12   
6553627      2136  4154817            b      2018-06-12           19:54   
6553929      2136  4662280            e      2018-07-06           23:24   
6553962      2136  4662301            b      2018-07-08           22:28   
6553991      2136  4661029            d      2018-07-11           00:50   
6554201      2136  4662438            b      2018-08-14           22:14   
6554351      2136  4662523            d      2018-08-28           23:30   
6554555      2136  4661958            a      2018-09-15           01:01   
6554567      2136  4662645            c      2018-09-15           22:37   
6554583      2136  4662656            a      2018-09-16           01:39   
6554883      2136  4662008            d      2018-10-04           21:35   
6554903      2136  4662847            d      2018-10-06           23:47   
6555028      2136   851436            h      2018-10-17           22:44   
6555308      2136  4663091            b      2018-11-11           23:40   
6555344      2136  4663117            c      2018-11-17           02:18   
6555394      2136  4662561            e      2018-11-20           20:18   
6555440      2136  4663170            d      2018-11-24           21:09   
6555645      2136   849905        

<b>[확인 된 내용]</b><br>
음수값은 매출이 아니라 취소금액이므로 row 에서 드랍해야 한다.

### 1.2.3 Null Data 확인

In [7]:
# 데이터 형태를 출력해 column, row 개수 확인
train.shape

(6556613, 9)

In [8]:
# 데이터 타입 확인 (불필요하게 큰 데이터는 사이즈를 줄여줄 수 있도록)
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6556613 entries, 0 to 6556612
Data columns (total 9 columns):
store_id            int64
card_id             int64
card_company        object
transacted_date     object
transacted_time     object
installment_term    int64
region              object
type_of_business    object
amount              float64
dtypes: float64(1), int64(3), object(5)
memory usage: 450.2+ MB


In [9]:
# 빈 데이터 개수 확인
train.isnull().sum() # 지역과 비즈니스 타입은 null 데이터가 많아 분석에 도움이 안 될 것으로 판단되므로 나중에 drop 을 고민하도록 한다.

store_id                  0
card_id                   0
card_company              0
transacted_date           0
transacted_time           0
installment_term          0
region              2042766
type_of_business    3952609
amount                    0
dtype: int64

In [10]:
# 지역 컬럼 결착율
2042766/6556613*100

31.155811697289437

In [11]:
# 업종 컬럼 결착율
3952609/6556613*100

60.28431142725672

### 1.3 확인한 값들을 이용한 처리방안
1. 데이터 개수는 6556613 개다 (매출액 별 기록이기 때문에 매장별 매출로 변환이 필요하다)
2. 매출액('amount')에 음수가 있다. (취소 처리 내역)
 - 매출액이 아니므로 데이터에서 드랍한다.
3. transacted_date, transacted_time 컬럼을 합쳐 시계열 데이터 처리를 지원해줄 수 있다.
4. 지역('region')과 업종 구분('type_of_business')은 결착율이 각가 31% 60% 이므로 변별력이 떨어진다. (드랍한다)
5. 주식지수 예측과 비슷한 단순 이동 평균 모델을 사용해 보도록 한다.
 - 월별 매출액 산출이 필요하다

# 2. 전처리

## 2.1 메모리 및 연산속도 향상을 위한 데이터 변형
1. 컬럼별 최소 최대값 확인 후 데이터 타입 변경 (메모리 사용량 최적화)

In [12]:
# int 타입인 컬럼만 최소 최대값 출력
for column_key in train.columns : # 컬럼 이름 조회
    if True == np.issubdtype( train[column_key].dtypes, np.integer ) : # 컬럼 별 데이터 타입 조회
        print( 'key: {}\n\tmin: {}\n\tmax: {}'.format( column_key, get_min(train, column_key), get_max(train, column_key) ))

key: store_id
	min: 0
	max: 2136
key: card_id
	min: 0
	max: 4663856
key: installment_term
	min: 0
	max: 93


<b>[데이터 변형을 위한 분석 결과]</b><br>
1. 컬럼 중 'store_id', 'card_id', 'installment_term' 만 정수형 변수이며 int32 최대값 (2,147,483,647) 을 넘지 않기 때문에 모든 데이터 크기를 축소해준다.<br>

In [13]:
train = train.astype( {'store_id': 'int16', 'card_id': 'int32', 'installment_term': 'int16'} )
train

store_id  card_id card_company transacted_date transacted_time  \
0               0        0            b      2016-06-01           13:13   
1               0        1            h      2016-06-01           18:12   
2               0        2            c      2016-06-01           18:52   
3               0        3            a      2016-06-01           20:22   
4               0        4            c      2016-06-02           11:06   
5               0        5            c      2016-06-02           13:09   
6               0        6            f      2016-06-02           15:33   
7               0        7            a      2016-06-02           17:18   
8               0        8            c      2016-06-02           18:30   
9               0        9            a      2016-06-02           19:56   
10              0       10            f      2016-06-02           22:26   
11              0       11            c      2016-06-03           12:43   
12              0       12            b      2016-06-03           19:05   
13              0       13            e      2016-06-03           21:46   
14              0       14            a      2016-06-04           11:15   
15              0       15            c      2016-06-04           12:58   
16              0       16            h      2016-06-04           13:42   
17              0       17            a      2016-06-04           14:32   
18              0       18            a      2016-06-04           14:53   
19              0       19            b      2016-06-04           19:35   
20              0       20            c      2016-06-04           20:30   
21              0       21            f      2016-06-05           11:09   
22              0       22            c      2016-06-05           11:31   
23              0       23            b      2016-06-05           15:09   
24              0       24            a      2016-06-06           11:35   
25              0       25            h      2016-06-06           17:55   
26              0       26            c      2016-06-06           18:14   
27              0       27            a      2016-06-06           18:53   
28              0       28            c      2016-06-09           10:43   
29              0       29            b      2016-06-09           12:21   
...           ...      ...          ...             ...             ...   
6556583      2136  4663842            a      2019-02-26           00:02   
6556584      2136  4663793            a      2019-02-26           00:29   
6556585      2136  4662280            e      2019-02-26           20:10   
6556586      2136  4663392            d      2019-02-26           23:01   
6556587      2136   795775            a      2019-02-26           23:08   
6556588      2136  4662603            f      2019-02-27           00:20   
6556589      2136  4663843            c      2019-02-27           01:04   
6556590      2136  4663844            b      2019-02-27           01:19   
6556591      2136  4663845            b      2019-02-27           02:15   
6556592      2136  1977050            d      2019-02-27           03:45   
6556593      2136  4663846            a      2019-02-27           21:38   
6556594      2136  4663847            g      2019-02-27           21:55   
6556595      2136  4663848            d      2019-02-27           22:10   
6556596      2136  4663706            d      2019-02-27           23:09   
6556597      2136  4663849            d      2019-02-27           23:20   
6556598      2136  4661654            a      2019-02-28           01:04   
6556599      2136  4663850            a      2019-02-28           02:35   
6556600      2136  4663107            a      2019-02-28           21:20   
6556601      2136  1980399            d      2019-02-28           22:08   
6556602      2136  4663851            b      2019-02-28           22:31   
6556603      2136  4663852            a      2019-02-28           22:35   
6556604      2136  4663853        

## 2.2 데이터 드랍
### 2.2.1 결착율 높은 데이터 드랍

In [14]:
# 'region'과 'type_of_business' 드랍
train = train.drop( 'region', axis=1)
train = train.drop( 'type_of_business', axis=1 )
train

store_id  card_id card_company transacted_date transacted_time  \
0               0        0            b      2016-06-01           13:13   
1               0        1            h      2016-06-01           18:12   
2               0        2            c      2016-06-01           18:52   
3               0        3            a      2016-06-01           20:22   
4               0        4            c      2016-06-02           11:06   
5               0        5            c      2016-06-02           13:09   
6               0        6            f      2016-06-02           15:33   
7               0        7            a      2016-06-02           17:18   
8               0        8            c      2016-06-02           18:30   
9               0        9            a      2016-06-02           19:56   
10              0       10            f      2016-06-02           22:26   
11              0       11            c      2016-06-03           12:43   
12              0       12            b      2016-06-03           19:05   
13              0       13            e      2016-06-03           21:46   
14              0       14            a      2016-06-04           11:15   
15              0       15            c      2016-06-04           12:58   
16              0       16            h      2016-06-04           13:42   
17              0       17            a      2016-06-04           14:32   
18              0       18            a      2016-06-04           14:53   
19              0       19            b      2016-06-04           19:35   
20              0       20            c      2016-06-04           20:30   
21              0       21            f      2016-06-05           11:09   
22              0       22            c      2016-06-05           11:31   
23              0       23            b      2016-06-05           15:09   
24              0       24            a      2016-06-06           11:35   
25              0       25            h      2016-06-06           17:55   
26              0       26            c      2016-06-06           18:14   
27              0       27            a      2016-06-06           18:53   
28              0       28            c      2016-06-09           10:43   
29              0       29            b      2016-06-09           12:21   
...           ...      ...          ...             ...             ...   
6556583      2136  4663842            a      2019-02-26           00:02   
6556584      2136  4663793            a      2019-02-26           00:29   
6556585      2136  4662280            e      2019-02-26           20:10   
6556586      2136  4663392            d      2019-02-26           23:01   
6556587      2136   795775            a      2019-02-26           23:08   
6556588      2136  4662603            f      2019-02-27           00:20   
6556589      2136  4663843            c      2019-02-27           01:04   
6556590      2136  4663844            b      2019-02-27           01:19   
6556591      2136  4663845            b      2019-02-27           02:15   
6556592      2136  1977050            d      2019-02-27           03:45   
6556593      2136  4663846            a      2019-02-27           21:38   
6556594      2136  4663847            g      2019-02-27           21:55   
6556595      2136  4663848            d      2019-02-27           22:10   
6556596      2136  4663706            d      2019-02-27           23:09   
6556597      2136  4663849            d      2019-02-27           23:20   
6556598      2136  4661654            a      2019-02-28           01:04   
6556599      2136  4663850            a      2019-02-28           02:35   
6556600      2136  4663107            a      2019-02-28           21:20   
6556601      2136  1980399            d      2019-02-28           22:08   
6556602      2136  4663851            b      2019-02-28           22:31   
6556603      2136  4663852            a      2019-02-28           22:35   
6556604      2136  4663853        

### 2.2.2 취소 데이터(매출액이 음수인 값) 드랍

In [15]:
train.drop(train.loc[train['amount']<0].index, inplace=True)
train

store_id  card_id card_company transacted_date transacted_time  \
0               0        0            b      2016-06-01           13:13   
1               0        1            h      2016-06-01           18:12   
2               0        2            c      2016-06-01           18:52   
3               0        3            a      2016-06-01           20:22   
4               0        4            c      2016-06-02           11:06   
5               0        5            c      2016-06-02           13:09   
6               0        6            f      2016-06-02           15:33   
7               0        7            a      2016-06-02           17:18   
8               0        8            c      2016-06-02           18:30   
9               0        9            a      2016-06-02           19:56   
10              0       10            f      2016-06-02           22:26   
11              0       11            c      2016-06-03           12:43   
12              0       12            b      2016-06-03           19:05   
13              0       13            e      2016-06-03           21:46   
14              0       14            a      2016-06-04           11:15   
15              0       15            c      2016-06-04           12:58   
16              0       16            h      2016-06-04           13:42   
17              0       17            a      2016-06-04           14:32   
18              0       18            a      2016-06-04           14:53   
19              0       19            b      2016-06-04           19:35   
20              0       20            c      2016-06-04           20:30   
21              0       21            f      2016-06-05           11:09   
22              0       22            c      2016-06-05           11:31   
23              0       23            b      2016-06-05           15:09   
24              0       24            a      2016-06-06           11:35   
25              0       25            h      2016-06-06           17:55   
26              0       26            c      2016-06-06           18:14   
27              0       27            a      2016-06-06           18:53   
28              0       28            c      2016-06-09           10:43   
29              0       29            b      2016-06-09           12:21   
...           ...      ...          ...             ...             ...   
6556582      2136  4663841            b      2019-02-26           00:00   
6556583      2136  4663842            a      2019-02-26           00:02   
6556584      2136  4663793            a      2019-02-26           00:29   
6556585      2136  4662280            e      2019-02-26           20:10   
6556586      2136  4663392            d      2019-02-26           23:01   
6556587      2136   795775            a      2019-02-26           23:08   
6556588      2136  4662603            f      2019-02-27           00:20   
6556589      2136  4663843            c      2019-02-27           01:04   
6556590      2136  4663844            b      2019-02-27           01:19   
6556591      2136  4663845            b      2019-02-27           02:15   
6556592      2136  1977050            d      2019-02-27           03:45   
6556593      2136  4663846            a      2019-02-27           21:38   
6556594      2136  4663847            g      2019-02-27           21:55   
6556595      2136  4663848            d      2019-02-27           22:10   
6556596      2136  4663706            d      2019-02-27           23:09   
6556597      2136  4663849            d      2019-02-27           23:20   
6556598      2136  4661654            a      2019-02-28           01:04   
6556599      2136  4663850            a      2019-02-28           02:35   
6556600      2136  4663107            a      2019-02-28           21:20   
6556601      2136  1980399            d      2019-02-28           22:08   
6556602      2136  4663851            b      2019-02-28           22:31   
6556603      2136  4663852        

### 2.2.3 시계열 데이터 (날짜와 시간)을 합쳐준다.

In [16]:
# 날짜 합치기
train['date'] = train['transacted_date']+' '+train['transacted_time']
train['date'] = pd.to_datetime(train['date'])

train.head(5)

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   
3         0        3            a      2016-06-01           20:22   
4         0        4            c      2016-06-02           11:06   

   installment_term       amount                date  
0                 0  1857.142857 2016-06-01 13:13:00  
1                 0   857.142857 2016-06-01 18:12:00  
2                 0  2000.000000 2016-06-01 18:52:00  
3                 0  7857.142857 2016-06-01 20:22:00  
4                 0  2000.000000 2016-06-02 11:06:00

In [17]:
# transacted_date 와 transacted_time 을 드랍한다.
train = train.drop( 'transacted_date', axis=1)
train = train.drop( 'transacted_time', axis=1 )
train

store_id  card_id card_company  installment_term        amount  \
0               0        0            b                 0   1857.142857   
1               0        1            h                 0    857.142857   
2               0        2            c                 0   2000.000000   
3               0        3            a                 0   7857.142857   
4               0        4            c                 0   2000.000000   
5               0        5            c                 0   2000.000000   
6               0        6            f                 0   2000.000000   
7               0        7            a                 0   7857.142857   
8               0        8            c                 0   2000.000000   
9               0        9            a                 0   1857.142857   
10              0       10            f                 0  22857.142857   
11              0       11            c                 0   2000.000000   
12              0       12            b                 0   1428.571429   
13              0       13            e                 0  14714.285714   
14              0       14            a                 0   2000.000000   
15              0       15            c                 0   3571.428571   
16              0       16            h                 0   1428.571429   
17              0       17            a                 0   2000.000000   
18              0       18            a                 0  14285.714286   
19              0       19            b                 0   6428.571429   
20              0       20            c                 0   2000.000000   
21              0       21            f                 0   1285.714286   
22              0       22            c                 0   2000.000000   
23              0       23            b                 0   7142.857143   
24              0       24            a                 0   7142.857143   
25              0       25            h                 0   2000.000000   
26              0       26            c                 0   6428.571429   
27              0       27            a                 0   1714.285714   
28              0       28            c                 0   2000.000000   
29              0       29            b                 0   2142.857143   
...           ...      ...          ...               ...           ...   
6556582      2136  4663841            b                 0   4428.571429   
6556583      2136  4663842            a                 0   6428.571429   
6556584      2136  4663793            a                 0   8142.857143   
6556585      2136  4662280            e                 0   4000.000000   
6556586      2136  4663392            d                 0  16571.428571   
6556587      2136   795775            a                 0   7571.428571   
6556588      2136  4662603            f                 0   7928.571429   
6556589      2136  4663843            c                 0   4500.000000   
6556590      2136  4663844            b                 0   6000.000000   
6556591      2136  4663845            b                 0   5142.857143   
6556592      2136  1977050            d                 0   6428.571429   
6556593      2136  4663846            a                 0   7857.142857   
6556594      2136  4663847            g                 0   7285.714286   
6556595      2136  4663848            d                 0   4000.000000   
6556596      2136  4663706            d                 0   5500.000000   
6556597      2136  4663849            d                 0  10428.571429   
6556598      2136  4661654            a                 0   8571.428571   
6556599      2136  4663850            a                 0   7428.571429   
6556600      2136  4663107            a                 0   3285.714286   
6556601      2136  1980399            d                 0   8714.285714   
6556602      2136  4663851            b                 0   3285.714286   
6556603      2136  4663852        

In [18]:
# 시간별 매출로 정렬하기 위해 date 를 index 로 지정한다.
train.set_index('date', inplace=True)

train

store_id  card_id card_company  installment_term  \
date                                                                    
2016-06-01 13:13:00         0        0            b                 0   
2016-06-01 18:12:00         0        1            h                 0   
2016-06-01 18:52:00         0        2            c                 0   
2016-06-01 20:22:00         0        3            a                 0   
2016-06-02 11:06:00         0        4            c                 0   
2016-06-02 13:09:00         0        5            c                 0   
2016-06-02 15:33:00         0        6            f                 0   
2016-06-02 17:18:00         0        7            a                 0   
2016-06-02 18:30:00         0        8            c                 0   
2016-06-02 19:56:00         0        9            a                 0   
2016-06-02 22:26:00         0       10            f                 0   
2016-06-03 12:43:00         0       11            c                 0   
2016-06-03 19:05:00         0       12            b                 0   
2016-06-03 21:46:00         0       13            e                 0   
2016-06-04 11:15:00         0       14            a                 0   
2016-06-04 12:58:00         0       15            c                 0   
2016-06-04 13:42:00         0       16            h                 0   
2016-06-04 14:32:00         0       17            a                 0   
2016-06-04 14:53:00         0       18            a                 0   
2016-06-04 19:35:00         0       19            b                 0   
2016-06-04 20:30:00         0       20            c                 0   
2016-06-05 11:09:00         0       21            f                 0   
2016-06-05 11:31:00         0       22            c                 0   
2016-06-05 15:09:00         0       23            b                 0   
2016-06-06 11:35:00         0       24            a                 0   
2016-06-06 17:55:00         0       25            h                 0   
2016-06-06 18:14:00         0       26            c                 0   
2016-06-06 18:53:00         0       27            a                 0   
2016-06-09 10:43:00         0       28            c                 0   
2016-06-09 12:21:00         0       29            b                 0   
...                       ...      ...          ...               ...   
2019-02-26 00:00:00      2136  4663841            b                 0   
2019-02-26 00:02:00      2136  4663842            a                 0   
2019-02-26 00:29:00      2136  4663793            a                 0   
2019-02-26 20:10:00      2136  4662280            e                 0   
2019-02-26 23:01:00      2136  4663392            d                 0   
2019-02-26 23:08:00      2136   795775            a                 0   
2019-02-27 00:20:00      2136  4662603            f                 0   
2019-02-27 01:04:00      2136  4663843            c                 0   
2019-02-27 01:19:00      2136  4663844            b                 0   
2019-02-27 02:15:00      2136  4663845            b                 0   
2019-02-27 03:45:00      2136  1977050            d                 0   
2019-02-27 21:38:00      2136  4663846            a                 0   
2019-02-27 21:55:00      2136  4663847            g                 0   
2019-02-27 22:10:00      2136  4663848            d                 0   
2019-02-27 23:09:00      2136  4663706            d                 0   
2019-02-27 23:20:00      2136  4663849            d                 0   
2019-02-28 01:04:00      2136  4661654            a                 0   
2019-02-28 02:35:00      2136  4663850            a                 0   
2019-02-28 21:20:00      2136  4663107            a                 0   
2019-02-28 22:08:00      2136  1980399            d                 0   
2019-02-28 22:31:00      2136  4663851            b                 0   
2019-02-28 22:35:00      2136  4663852            a                 0   
2019-0

In [19]:
train.index

DatetimeIndex(['2016-06-01 13:13:00', '2016-06-01 18:12:00',
               '2016-06-01 18:52:00', '2016-06-01 20:22:00',
               '2016-06-02 11:06:00', '2016-06-02 13:09:00',
               '2016-06-02 15:33:00', '2016-06-02 17:18:00',
               '2016-06-02 18:30:00', '2016-06-02 19:56:00',
               ...
               '2019-02-28 22:31:00', '2019-02-28 22:35:00',
               '2019-02-28 22:43:00', '2019-02-28 22:45:00',
               '2019-02-28 23:03:00', '2019-02-28 23:20:00',
               '2019-02-28 23:24:00', '2019-02-28 23:24:00',
               '2019-02-28 23:27:00', '2019-02-28 23:54:00'],
              dtype='datetime64[ns]', name='date', length=6483513, freq=None)

### 2.2.4 월 매출액으로 변환

In [20]:
df_month = pd.DataFrame()
for i in tqdm( train.store_id.unique() ): #매장별 매출액 합산
    data_num = train[train.store_id==i]
    sum_amount = data_num['amount'].resample(rule='m').sum()
    data_mon = pd.concat([sum_amount],axis=1)
    data_mon.insert(0,'store_id',i)
    df_month= pd.concat([df_month,data_mon],axis=0)

df_month

100%|██████████| 1967/1967 [00:25<00:00, 76.63it/s]


store_id        amount
date                              
2016-06-30         0  7.555714e+05
2016-07-31         0  1.005000e+06
2016-08-31         0  8.734286e+05
2016-09-30         0  8.978571e+05
2016-10-31         0  8.374286e+05
2016-11-30         0  6.971429e+05
2016-12-31         0  7.618571e+05
2017-01-31         0  5.856429e+05
2017-02-28         0  7.940000e+05
2017-03-31         0  7.202571e+05
2017-04-30         0  6.852857e+05
2017-05-31         0  7.444286e+05
2017-06-30         0  6.820000e+05
2017-07-31         0  7.465714e+05
2017-08-31         0  7.490000e+05
2017-09-30         0  8.408571e+05
2017-10-31         0  6.005714e+05
2017-11-30         0  6.308571e+05
2017-12-31         0  8.198571e+05
2018-01-31         0  6.431614e+05
2018-02-28         0  6.852857e+05
2018-03-31         0  8.484286e+05
2018-04-30         0  6.361429e+05
2018-05-31         0  6.884286e+05
2018-06-30         0  7.072857e+05
2018-07-31         0  7.637143e+05
2018-08-31         0  6.818571e+05
2018-09-30         0  6.552857e+05
2018-10-31         0  7.390000e+05
2018-11-30         0  6.760000e+05
...              ...           ...
2018-07-31      2135  4.384286e+05
2018-08-31      2135  5.661429e+05
2018-09-30      2135  5.098571e+05
2018-10-31      2135  8.504286e+05
2018-11-30      2135  5.894286e+05
2018-12-31      2135  5.544286e+05
2019-01-31      2135  4.622857e+05
2019-02-28      2135  4.042857e+05
2017-05-31      2136  5.036429e+05
2017-06-30      2136  2.722886e+06
2017-07-31      2136  2.989500e+06
2017-08-31      2136  3.379929e+06
2017-09-30      2136  3.190643e+06
2017-10-31      2136  3.373429e+06
2017-11-30      2136  2.711714e+06
2017-12-31      2136  2.884000e+06
2018-01-31      2136  1.882286e+06
2018-02-28      2136  2.031929e+06
2018-03-31      2136  2.356786e+06
2018-04-30      2136  2.197429e+06
2018-05-31      2136  2.202571e+06
2018-06-30      2136  2.342286e+06
2018-07-31      2136  1.614643e+06
2018-08-31      2136  1.234300e+06
2018-09-30      2136  2.789857e+06
2018-10-31      2136  2.026857e+06
2018-11-30      2136  2.170500e+06
2018-12-31      2136  2.452500e+06
2019-01-31      2136  1.879500e+06
2019-02-28      2136  2.266429e+06

[60913 rows x 2 columns]

# 3. 모델링 및 예측

In [21]:
# 예측 모델을 아직 직접 구현하긴 어려우니 참조하기로 한다.
# 단 코드 분석은 해보도록 한다.
# https://www.dacon.io/competitions/official/140472/codeshare/1734?page=1&dtype=recent&ptype=pub
def sub(df,n):
    # 3달치 empty DataFrame 생성
    concat_3mon = pd.DataFrame(index=pd.to_datetime(['2019-03-31', '2019-04-30', '2019-05-31']))
    df_sma=pd.DataFrame(columns=['store_id','amount'])

    for i in df_month.store_id.unique():
        # store_id 별 amount가 담긴 sf_set 데임터 프레임을 생성
        df_set = pd.DataFrame(df[df.store_id == i].amount)
        # store_id 별 df_set에 contact_3mon 을 병합한다.
        sma_train = pd.concat([df_set,concat_3mon], axis=0)
        num = df_set.amount.rolling(window=n,min_periods=1).mean().values[-3:].sum()
        df_sma.loc[i,'store_id']=i
        df_sma.loc[i,'amount']=num
        
    return df_sma

In [22]:
#월별 예측 매출액 저장
for i in tqdm(range(4, 7)):
    sma_sub = sub(df_month, i)
    sma_sub.to_csv(f'data/funda_{i}sma_sub.csv',index=False)

100%|██████████| 3/3 [00:22<00:00,  7.41s/it]


In [25]:
# 저장된 값을 로드해 하나의 파일로 합쳐준다.
list_files = glob.glob('data/*')
df_all = []
for file in list_files :
    df = pd.read_csv(file)
    df_all.append(df)
    
df_combine = pd.concat( df_all, axis=0, ignore_index=True)
df_combine.to_csv('output/output.csv', index=False)